engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java

```diff
@@ -627,6 +628,182 @@ public void shouldResolveIncidentDueToCompletionCondition() {
         .isTrue();
   }
 
+  /**
+   * This test covers the scenario where a multi instance cannot be completed, because updating the
+   * output collection fails due to an index out of bounds. The index out of bounds is caused
+   * because a) the output collection is initialized with the cardinality of the multi instance when
+   * the multi instance is activated, and b) the collection is modified and shrunk to a smaller
+   * size.
+   */
+  @Test // regression test for #9143
+  public void shouldCreateIncidentIfOutputElementCannotBeReplacedInOutputCollection() {
+    // given
+    final var processId = "index-out-of-bounds-in-output-collection";
+    final var collectionWithThreeElements = "=[1,2,3]";
+    final var collectionWithNoElements = "=[]";
+    final var outputCollectionName = "outputItems";
+
+    final var process =
+        createProcessThatModifiesOutputCollection(
+            processId, collectionWithThreeElements, collectionWithNoElements, outputCollectionName);
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+
+    // when
+    final var processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(processId).create();
+
+    // then
+    final Record<IncidentRecordValue> incidentEvent =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(incidentEvent.getValue())
+        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
+        .hasErrorMessage(
+            "Unable to update item in output collection 'outputItems' at position 1 because the size of the collection is: 0. This happens when multiple BPMN elements write to the same variable.")
+        .hasProcessInstanceKey(processInstanceKey);
+  }
+
+  @Test
+  public void shouldResolveIncidentRaisedByOutputElementCannotBeReplacedInOutputCollection() {
```

Comment:
:wrench: We could merge this test case into the previous test case `shouldCreateIncidentIfOutputElementCannotBeReplacedInOutputCollection`.

I know that this is handled differently until now. But both test cases have much in common. We would reduce the duplicated code and shrink the test class by merging both test cases.

Same with the other test.

Reply:
Not sure what you have in mind, tbh.